src/system_monitor_pg.erl

```diff
@@ -0,0 +1,264 @@
+start() ->
+  {ok, _Pid} = gen_server:start_link({local, ?SERVER}, ?MODULE, [], []),
+  ok.
+
+stop() ->
+  gen_server:stop(?SERVER).
+
+produce(Type, Events) ->
+  gen_server:cast(?SERVER, {produce, Type, Events}).
+
+%%%_* Callbacks =================================================================
+init(_Args) ->
+  erlang:process_flag(trap_exit, true),
+  {ok, #{}, {continue, start_pg}}.
+
+handle_continue(start_pg, State) ->
+  Conn = initialize(),
+  case Conn of
+    undefined ->
+      timer:send_after(?FIVE_SECONDS, reinitialize);
+    Conn ->
+      ok
+  end,
+  timer:send_after(?ONE_HOUR, mk_partitions),
+  {noreply, State#{connection => Conn, buffer => buffer_new()}}.
+
+handle_call(_Msg, _From, State) ->
+  {reply, ok, State}.
+
+handle_info({'EXIT', Conn, _Reason}, #{connection := Conn} = State) ->
+  timer:send_after(?FIVE_SECONDS, reinitialize),
+  {noreply, State};
+handle_info({'EXIT', _Conn, _Reason}, #{connection := undefined} = State) ->
+  timer:send_after(?FIVE_SECONDS, reinitialize),
+  {noreply, State};
+handle_info({'EXIT', _Conn, normal}, State) ->
+  {noreply, State};
+handle_info(mk_partitions, #{connection := undefined} = State) ->
+  timer:send_after(?ONE_HOUR, mk_partitions),
+  {noreply, State};
+handle_info(mk_partitions, #{connection := Conn} = State) ->
+  mk_partitions(Conn),
+  timer:send_after(?ONE_HOUR, mk_partitions),
+  {noreply, State};
+handle_info(reinitialize, State) ->
+  {noreply, State#{connection => initialize()}}.
+
+handle_cast({produce, Type, Events}, #{connection := undefined, buffer := Buffer} = State) ->
+  {noreply, State#{buffer => buffer_add(Buffer, {Type, Events})}};
+handle_cast({produce, Type, Events}, #{connection := Conn, buffer := Buffer} = State) ->
+  MaxMsgQueueSize = application:get_env(?APP, max_message_queue_len, 1000),
+  case process_info(self(), message_queue_len) of
+    {_, N} when N > MaxMsgQueueSize ->
+      {noreply, State};
+    _ ->
+      lists:foreach(fun({Type0, Events0}) ->
+                      lists:foreach(fun(Event) ->
+                                      {ok, _} = epgsql:equery(Conn,
+                                                              query(Type0),
+                                                              params(Type0, Event))
+                                    end, Events0)
+                    end, buffer_add(Buffer, {Type, Events})),
+      {noreply, State#{buffer => buffer_new()}}
+  end.
+
+format_status(normal, [_PDict, State]) ->
+  [{data, [{"State", State}]}];
+format_status(terminate, [_PDict, State]) ->
+  State#{buffer => buffer_new()}.
+
+terminate(_Reason, #{connection := Conn}) ->
+  epgsql:close(Conn).
+
+%%%_* Internal buffer functions ===============================================
+buffer_new() ->
+  [].
+
+buffer_add(Buffer, Element) ->
+  MaxBufferSize = application:get_env(?APP, max_buffer_size, 1000),
+  case buffer_size(Buffer) > MaxBufferSize of
+    true -> [Element | lists:reverse(tl(lists:reverse(Buffer)))];
```

Comment:
O(N^2) here. I think the buffer needs a bit more structure, but I'll need to think about the details for a while.

Reply:
We have talked about moving to queue.erl, but the length function will still be dominating. `{Size::non_neg_integer(), queue()}` could be a solution if we are concerned with the run time of this. Introducing such a data type would need some good testing.

Reply:
Implemented it as a `{Len, queue()}` now